src/library/Connect/LocalEnvironmentManager.cs

```diff
@@ -421,67 +421,69 @@ private void _ReportProgress(EventLevel eventLevel, string message, params objec
+
+                var host = _useKubernetesServiceEnvironmentVariables || string.Equals(endpoint.DnsName, DAPR, StringComparison.OrdinalIgnoreCase)
+                    ? endpoint.LocalIP.ToString()
+                    : endpoint.DnsName;
+
+                if (string.Equals(serviceName, "KUBERNETES", StringComparison.OrdinalIgnoreCase))
+                    host = _kubernetesClient.HostName;
+
+                var unnamedPort = _useKubernetesServiceEnvironmentVariables || string.Equals(endpoint.DnsName, DAPR, StringComparison.OrdinalIgnoreCase)
+                    ? endpoint.Ports[0].LocalPort.ToString()
+                    : endpoint.Ports[0].RemotePort.ToString();
+
+                // Service Host
+                result[$"{serviceName}_SERVICE_HOST"] = host;
+
+                // Service Port: allocate the first port in the service to the backwards-compatible environment variable in keeping with Kubernetes source code
+                result[$"{serviceName}_SERVICE_PORT"] = unnamedPort;
+                result[$"{serviceName}_PORT"] = $"{endpoint.Ports[0].Protocol}://{host}:{unnamedPort}";
+
+                // All named ports (only the first may be unnamed according to Kubernetes source code)
                 foreach (var portPair in endpoint.Ports)
                 {
-                    this._AddKubernetesServiceEnv(result,
-                                                  endpoint.DnsName,
-                                                  (this._useKubernetesServiceEnvironmentVariables || StringComparer.OrdinalIgnoreCase.Equals(endpoint.DnsName, DAPR)) ? endpoint.LocalIP.ToString() : endpoint.DnsName,
-                                                  portPair.Protocol,
-                                                  (this._useKubernetesServiceEnvironmentVariables || StringComparer.OrdinalIgnoreCase.Equals(endpoint.DnsName, DAPR)) ? portPair.LocalPort : portPair.RemotePort);
-                }
-            }
-            return result;
-        }
+                    var port = _useKubernetesServiceEnvironmentVariables || string.Equals(endpoint.DnsName, DAPR, StringComparison.OrdinalIgnoreCase)
+                        ? portPair.LocalPort
+                        : portPair.RemotePort;
 
-        /// <summary>
-        /// Add Kubernetes environment variables to existing list of environment variables
-        /// </summary>
-        private void _AddKubernetesServiceEnv(IDictionary<string, string> envVariables, string dnsName, string host, string protocol, int port)
-        {
-            dnsName = dnsName.ToUpperInvariant();
-            // because we are using dns name instead of service we have to retrieve it by splitting when needed
-            // If this ever cause issues we should consider larger refactor where we add serviceName member varaible to EndpointInfo class.
-            string serviceName = dnsName.Split(".").First();
+                    var protocolUpper = portPair.Protocol.ToUpperInvariant();

```

Comment:
Before there was a check for this being empty and if so it would set it to tcp. Can we keep the check?

Reply:
Ah yes, I can see no existing code paths currently set this property directly and both constructors default it to the constant `KubernetesConstants.Protocols.Tcp` with no available empty constructor to allow for object initialization but it could be overwritten after object creation.

I will write a defensive check for this specific scenario here but I am concerned this would also need to be replicated in every place where the value of this property needs to be defaulted.

In the interest of DRY, could there be an argument for adding a default/empty constructor to allow object initialization and defaulting the property so this would become the ubiquitous default for this property, both steps are required to avoid a code smell:

```c# 
public PortPair() 
{
}

public string Protocol { get; set; } = KubernetesConstants.Protocols.Tcp;
```